cpp/src/reductions/reduction_dispatcher_multistep.cuh

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_REDUCTION_DISPATCHER_MULTISTEP_CUH
+#define CUDF_REDUCTION_DISPATCHER_MULTISTEP_CUH
+
+#include "reduction_functions.cuh"
+
+namespace cudf {
+namespace reductions {
+
+// Reduction for mean, var, std
+// It requires extra step after single step reduction call
+template<typename T_in, typename T_out, typename Op, bool has_nulls>
+void ReduceMultiStepOp(const gdf_column *input,
+                   gdf_scalar* scalar, gdf_size_type ddof, cudaStream_t stream)
+{
+    gdf_size_type valid_count = input->size - input->null_count;
+
+    T_out identity = Op::Op::template identity<T_out>();
+    using intermediateOp = typename Op::template Intermediate<T_out>;
+    // Itype: intermediate structure, output type of `reduction_op` and
+    // input type of `intermediateOp::ComputeResult`
+    using Itype = typename intermediateOp::IType;
+    Itype intermediate{0};
+
+    // allocate temporary memory for the dev_result
+    void *dev_result = NULL;
+    RMM_TRY(RMM_ALLOC(&dev_result, sizeof(Itype), stream));
+
+    // initialize output by identity value
+    CUDA_TRY(cudaMemcpyAsync(dev_result, &intermediate,
+            sizeof(Itype), cudaMemcpyHostToDevice, stream));
+    CHECK_STREAM(stream);
+
+    // reduction by iterator
+    auto it = intermediateOp::template make_iterator<has_nulls, T_in, T_out>(input, identity);
+    reduction_op(static_cast<Itype*>(dev_result), it, input->size, intermediate,
+        typename Op::Op{}, stream);
+
+    // read back the dev_result to host memory
+    // TODO: asynchronous copy
```

Comment:
Does this need to be a TODO? Can we just do it now? You already use an async memcpy to the device above.

Reply:
Yes, it is still needed since it read back from device memory.
The synchronization is the future task.

Reply:
Can you just add a stream sync after the kernel?

Reply:
We've already discussed about this TODO and stream sync before.
Generally, read back from device to host is heavy operation.
At that time, we discussed that we may need some synchronization method like `std::async` and 'd like to use asynchronous copy here.

Though this is out of topic, I feel gdf_device_scalar will be better for cudf::reduction result.

> Can you just add a stream sync after the kernel?

Before, we'd add CUDA_TRY(cudaStreamSynchronize(stream), however we removed it since `stream` = 0 here.